client-lite/src/trace/do_log.cpp

```diff
@@ -0,0 +1,437 @@
+        DoLogInfo("Logger stats: [creation_attempts: %u, created: %u, deleted: %u, threadproc: %lu, flush_writes: %lu, writes_f: %u], "
+            "buffer: [resets: %u, flushes: %u, writes_s: %u, writes_f: %u]",
+            _stats.numFileCreationAttempts, _stats.numFilesCreated, _stats.numFilesDeleted, _stats.numFlushThreadLoops,
+            _stats.numFlushWrites, _stats.numWritesFailed,
+            bufferStats.numResets, bufferStats.numFlushes, bufferStats.numWritesSucceeded, bufferStats.numWritesFailed);
+
+        std::unique_lock<std::mutex> lock(_mutex);
+        if (_logFile.is_open())
+        {
+            _writeBuffer.Flush(_logFile);
+            _logFile.close();
+        }
+    }
+
+    void WriteV(Level level, const char* pszFunc, unsigned int uLine, HRESULT hrIn, const char* pszFmt, va_list argList)
+    {
+        if (level > _maxLogLevel)
+        {
+            return;
+        }
+
+        // TBD save and restore errno?
+
+        // Make a copy of argList to allow its reuse
+        va_list argCopy;
+        va_copy(argCopy, argList);
+
+        std::unique_lock<std::mutex> lock(_mutex);
+
+        HRESULT hr = _writeBuffer.Write(_LevelToString(level), pszFunc, uLine, hrIn, pszFmt, argCopy);
+        if (hr == STRSAFE_E_INSUFFICIENT_BUFFER)
+        {
+            hr = _writeBuffer.Flush(_logFile);
+            DO_ASSERT(SUCCEEDED(hr));
+            if (SUCCEEDED(hr))
+            {
+                _RotateLogFilesIfNeeded();
+
+                va_copy(argCopy, argList);
+                hr = _writeBuffer.Write(_LevelToString(level), pszFunc, uLine, hrIn, pszFmt, argCopy);
+            }
+            ++_stats.numFlushWrites;
+        }
+
+        if (FAILED(hr))
+        {
+            ++_stats.numWritesFailed;
+        }
+    }
+};
+
+void g_LogRuntimeFailure(const docli::FailureInfo& failure) noexcept
+{
+    // Runtime failures, including caught exceptions, are logged here.
+    // By default, pszFunction is present in debug code, not retail.
+    // Similarly, pszCode (string version of macro contents) is debug only.
+    // pszMessage is only present when using a _MSG macro, such as LOG_IF_FAILED_MSG.
+    if (failure.pszMessage != nullptr)
+    {
+        WriteResult(Level::Error, failure.pszFunction, failure.uLineNumber, failure.hr, "%s [%s, %d]", failure.pszMessage, failure.pszFile,
+            failure.uLineNumber);
+    }
+    else if (failure.pszCode != nullptr)
+    {
+        WriteResult(Level::Error, failure.pszFunction, failure.uLineNumber, failure.hr, "%s [%s, %d]", failure.pszCode, failure.pszFile,
+            failure.uLineNumber);
+    }
+    else
+    {
+        WriteResult(Level::Error, failure.pszFunction, failure.uLineNumber, failure.hr, "[%s, %d]", failure.pszFile, failure.uLineNumber);
+    }
+}
+
+// LoggerImpl destructor performs non-trivial tasks like waiting for thread termination.
+// Use raw pointer to avoid executing the destructor on process termination if Close() was not invoked.
```

Comment:
This is to avoid hanging the process for the potential 3 min wait time correct? Don't we lose the last bit of logging if the process terminates early (i.e. manually for example with systemctl restart on the agent daemon?)

Reply:
You're right on the hanging part, although I don't know if that applies to linux (windows will definitely hang if we wait for a thread to terminate within the process termination flow). I'll see if I can simplify this later.

All normal process termination will invoke Close() and thus the LoggerImpl destructor will peform the final flush. So we'll lose data only if the process crashes or is killed externally.